flows/triggers/base_test.go

```diff
@@ -1,13 +1,244 @@
+					},
+					"uuid": "8d339613-f0be-48b7-92ee-155f4c7576f8"
+				},
+				"flow": {
+					"name": "Registration",
+					"uuid": "7c37d7e5-6468-4b31-8109-ced2ef8b5ddc"
+				},
+				"triggered_on": "2018-10-20T09:49:31.23456789Z",
+				"type": "campaign"
+			}`,
+		},
+		{
+			triggers.NewChannelTrigger(
+				env,
+				flow,
+				contact,
+				triggers.NewChannelEvent("new-conversation", channel),
+				types.NewEmptyXMap(),
+				triggeredOn,
+			),
+			`{
+				"contact": {
+					"created_on": "2018-10-18T14:20:30.000123456Z",
+					"id": 0,
+					"language": "eng",
+					"name": "Bob",
+					"urns": [],
+					"uuid": "c00e5d67-c275-4389-aded-7d8b151cbd5b"
+				},
+				"environment": {
+					"date_format": "YYYY-MM-DD",
+					"redaction_policy": "none",
+					"time_format": "tt:mm",
+					"timezone": "UTC"
+				},
+				"event": {
+					"channel": {
+						"name": "Facebook",
+						"uuid": "8cd472c4-bb85-459a-8c9a-c04708af799e"
+					},
+					"type": "new-conversation"
+				},
+				"flow": {
+					"name": "Registration",
+					"uuid": "7c37d7e5-6468-4b31-8109-ced2ef8b5ddc"
+				},
+				"params": {},
+				"triggered_on": "2018-10-20T09:49:31.23456789Z",
+				"type": "channel"
+			}`,
+		},
+		{
+			triggers.NewFlowActionTrigger(
+				env,
+				flow,
+				contact,
+				json.RawMessage(`{"uuid": "084e4bed-667c-425e-82f7-bdb625e6ec9e"}`),
+				triggeredOn,
+			),
+			`{
+				"contact": {
+					"created_on": "2018-10-18T14:20:30.000123456Z",
+					"id": 0,
+					"language": "eng",
+					"name": "Bob",
+					"urns": [],
+					"uuid": "c00e5d67-c275-4389-aded-7d8b151cbd5b"
+				},
+				"environment": {
+					"date_format": "YYYY-MM-DD",
+					"redaction_policy": "none",
+					"time_format": "tt:mm",
+					"timezone": "UTC"
+				},
+				"flow": {
+					"name": "Registration",
+					"uuid": "7c37d7e5-6468-4b31-8109-ced2ef8b5ddc"
+				},
+				"run_summary": {
+					"uuid": "084e4bed-667c-425e-82f7-bdb625e6ec9e"
+				},
+				"triggered_on": "2018-10-20T09:49:31.23456789Z",
+				"type": "campaign"
+			}`,
+		},
+		{
+			triggers.NewManualTrigger(
+				env,
+				flow,
+				contact,
+				types.NewXArray(types.NewXText("foo")),
+				triggeredOn,
+			),
+			`{
+				"contact": {
+					"created_on": "2018-10-18T14:20:30.000123456Z",
+					"id": 0,
+					"language": "eng",
+					"name": "Bob",
+					"urns": [],
+					"uuid": "c00e5d67-c275-4389-aded-7d8b151cbd5b"
+				},
+				"environment": {
+					"date_format": "YYYY-MM-DD",
+					"redaction_policy": "none",
+					"time_format": "tt:mm",
+					"timezone": "UTC"
+				},
+				"flow": {
+					"name": "Registration",
+					"uuid": "7c37d7e5-6468-4b31-8109-ced2ef8b5ddc"
+				},
+				"params": [
+					"foo"
+				],
+				"triggered_on": "2018-10-20T09:49:31.23456789Z",
+				"type": "manual"
+			}`,
+		},
+		{
+			triggers.NewMsgTrigger(
+				env,
+				flow,
+				contact,
+				flows.NewMsgIn(flows.MsgUUID("c8005ee3-4628-4d76-be66-906352cb1935"), urns.URN("tel:+1234567890"), channel, "Hi there", nil),
```

Comment:
@nicpottier id and externalID set via `SetID` and `SetExternalID` now to keep this constructor a bit simpler